README.md
# EatMySpam for WordPress

This plugin allows you to use the free EatMySpam public API to check your WordPress comments for spam.



CHANGELOG.md
# Changelog

* (0.8.1) minor language fixes
* (0.8.0) language detection: you can now define languages that are allowed or languages that are not allowed to be used in the comment system
* (0.8.0) garbage collection: you can now define a number of days after which spam comments are deleted
* (0.7.3) reduce timeout for eat my spam requests
* (0.7.2) clarify settings description and order
* (0.7.2) cleanup plugin options on uninstall
* (0.7.1) clarify description in readme.txt that plugin uses a web service to classify spam
* (0.7.1) clarify features in readme.txt
* (0.7.0) add first version of contact-form-7 integration
* (0.6.3) internal changes
* (0.6.2) bump version
* (0.6.1) fix typos in plugin metadata
* (0.6.0) add add more information to spam notification mail
* (0.6.0) check compatibility with PHP 7.0 and PHP 7.1
* (0.6.0) report ham and spam classification to EatMySpam api (can be disabled in settings)
* (0.6.0) do not include wp-admin links in notification mail, see https://core.trac.wordpress.org/ticket/40081
* (0.5.0) add possibility to translate plugin
* (0.5.0) add translations for de_DE
* (0.4.1) fix in_array warnings on settings page
* (0.4.0) Spam messages are tagged as spam be default, but there is an option to delete those messages directly on settings page.
* (0.4.0) New option to send spam notification emails to site admin if a message is considered spam.
* (0.3.1) Unify plugin name.
* (0.3.0) Add threshold parameter on settings page where custom spam threshold can be set.
* (0.2.0) Add settings page where rulesets can be excluded.
* (0.1.0) Initial Release.

eat-my-spam/views/settings.php
<div class="wrap">

    <h1><?php esc_html_e( 'EatMySpam Settings', 'eat-my-spam' ); ?></h1>

    <form method="post" action="options.php">

		<?php settings_fields( 'eat-my-spam-settings' ); ?>

		<?php do_settings_sections( 'eat-my-spam-settings' ); ?>

        <h2><?php esc_html_e( 'Settings:', 'eat-my-spam' ); ?></h2>

        <table>

            <tr>
                <td colspan="3">
                    <label for="eatmyspam_threshold"><?php esc_html_e( 'Spam threshold:', 'eat-my-spam' ); ?></label><br>
                    <input type="number" id="eatmyspam_threshold" name="eatmyspam_threshold"
                           value="<?php echo get_option( 'eatmyspam_threshold', 5 ); ?>">
                </td>
            </tr>


            <tr>
                <td colspan="3">
                    <input id="eatmyspam_send_notifications" name="eatmyspam_send_notifications"
                           type="checkbox" <?php echo ( get_option( 'eatmyspam_send_notifications' ) === 'on' ) ? 'checked="checked"' : ''; ?>/>
                    <label for="eatmyspam_send_notifications"><?php esc_html_e( 'Send spam notifications to admin?', 'eat-my-spam' ); ?></label>
                </td>
            </tr>

            <tr>
                <td colspan="3">
                    <input id="eatmyspam_disable_reports" name="eatmyspam_disable_reports"
                           type="checkbox" <?php echo ( get_option( 'eatmyspam_disable_reports' ) === 'on' ) ? 'checked="checked"' : ''; ?>/>
                    <label for="eatmyspam_disable_reports"><?php esc_html_e( 'Disable automatic reports to EatMySpam server?', 'eat-my-spam' ); ?></label>
                </td>
            </tr>

            <tr>
                <td colspan="3">
                    <input id="eatmyspam_disable_cf7_integration" name="eatmyspam_disable_cf7_integration"
                           type="checkbox" <?php echo ( get_option( 'eatmyspam_disable_cf7_integration' ) === 'on' ) ? 'checked="checked"' : ''; ?>/>
                    <label for="eatmyspam_disable_cf7_integration"><?php esc_html_e( 'Disable Contact Form 7 integration?', 'eat-my-spam' ); ?></label>
                </td>
            </tr>

            <tr>
                <td colspan="3">
                    <input id="eatmyspam_remove_spam" name="eatmyspam_remove_spam"
                           type="checkbox" <?php echo ( get_option( 'eatmyspam_remove_spam' ) === 'on' ) ? 'checked="checked"' : ''; ?>/>
                    <label for="eatmyspam_remove_spam"><?php esc_html_e( 'Remove spam directly?', 'eat-my-spam' ); ?>
                        <span class="color: grey;">(<?php esc_html_e( 'Only works with WordPress comments. Otherwise spam is just tagged as spam and can be reviewed in the comment management area.', 'eat-my-spam' ); ?>)</span>
                    </label>
                </td>
            </tr>

            <tr>
                <td colspan="3">
                    <label for="eatmyspam_delete_spam_after_days"><?php esc_html_e( 'Remove spam after amount of days:', 'eat-my-spam' ); ?>
                        <span class="color: grey;">(<?php esc_html_e( 'Only works with WordPress comments. If you do not delete spam directly, spam will be deleted automatically after this configured number of days. Defaults to 0 which disables this function.', 'eat-my-spam' ); ?>)</span>
                    </label><br>
                    <input id="eatmyspam_delete_spam_after_days" name="eatmyspam_delete_spam_after_days"
                           type="number" value="<?php echo get_option( 'eatmyspam_delete_spam_after_days', 0 ); ?>"/>
                </td>
            </tr>
        </table>

        <h2><?php esc_html_e( 'Excluded rulesets:', 'eat-my-spam' ); ?></h2>

        <table>

            <tr>
                <th></th>
                <th><?php esc_html_e( 'Ruleset key', 'eat-my-spam' ); ?></th>
                <th><?php esc_html_e( 'Ruleset description', 'eat-my-spam' ); ?></th>
            </tr>

			<?php foreach ( $rulesets as $ruleset ) : ?>
                <tr>
                    <td><input id="eatmyspam_exclude_ruleset_<?php echo esc_attr( $ruleset->key ); ?>"
                               type="checkbox"
                               name="eatmyspam_excluded_rulesets[]" <?php echo in_array( $ruleset->key, $excludedRulesets ) ? 'checked="checked"' : ''; ?>
                               value="<?php echo esc_attr( $ruleset->key ); ?>"/>
                    </td>
                    <td>
                        <label for="eatmyspam_exclude_ruleset_<?php echo esc_attr( $ruleset->key ); ?>"><?php echo esc_html( $ruleset->key ); ?></label>
                    </td>
                    <td><?php echo esc_html( $ruleset->title ); ?></td>
                </tr>
			<?php endforeach; ?>

        </table>

        <h2><?php esc_html_e( 'Language Detection:', 'eat-my-spam' ); ?></h2>

        <table style="width: 100%;">

            <tr>
                <td colspan="3">
                    <label for="eatmyspam_allowed_languages"><?php esc_html_e( 'Allowed languages:', 'eat-my-spam' ); ?></label><br>
                    <select id="eatmyspam_allowed_languages" name="eatmyspam_allowed_languages[]" class="select2" multiple>
	                    <?php foreach ( $languages as $language ) : ?>
                        <option value="<?php echo esc_attr($language->{"2"}); ?>" <?php echo ( in_array( $language->{"2"}, $allowedLanguages ) ) ? 'selected="selected"' : ''; ?>>
                            <?php echo esc_html( $language->name ); ?>
                        </option>
                        <?php endforeach; ?>
                    </select>
                </td>
            </tr>

            <tr>
                <td colspan="3">
                    <label for="eatmyspam_disallowed_languages"><?php esc_html_e( 'Disallowed languages:', 'eat-my-spam' ); ?></label><br>
                    <select id="eatmyspam_disallowed_languages" name="eatmyspam_disallowed_languages[]" class="select2" multiple>
				        <?php foreach ( $languages as $language ) : ?>
                            <option value="<?php echo esc_attr($language->{"2"}); ?>" <?php echo ( in_array( $language->{"2"}, $disallowedLanguages ) ) ? 'selected="selected"' : ''; ?>>
						        <?php echo esc_html( $language->name ); ?>
                            </option>
				        <?php endforeach; ?>
                    </select>
                </td>
            </tr>

        </table>

		<?php submit_button(); ?>

    </form>

</div>

eat-my-spam/classes/class.eat-my-spam.php
<?php

/**
 * Security, checks if WordPress is running
 **/
if ( ! function_exists( 'add_action' ) ) {
	header( 'Status: 403 Forbidden' );
	header( 'HTTP/1.1 403 Forbidden' );
	exit();
}


/**
 * EatMySpam class
 *
 * @package EatMySpam
 * @author Daniel Jost
 **/
final class EatMySpam {

	const API_HOST = 'api.eat-my-spam.de';
	const VERSION = '0.6.2';

	/**
	 * Constructor
	 *
	 * @access public
	 * @author Daniel Jost
	 **/
	public function __construct() {
		add_filter( 'preprocess_comment', array( $this, 'check_comment' ), 1 );
	}

	/**
	 * check whether comment is spam
	 *
	 * @param array $comment
	 *
	 * @return array
	 */
	public function check_comment( $comment ) {

		$post              = get_post( $comment['comment_post_ID'] );
		$excluded_rulesets = get_option( 'eatmyspam_excluded_rulesets', array() );

		$result = $this->do_post( 'analyze', array(
			'message'          => $comment['comment_content'],
			'excludedRulesets' => $excluded_rulesets,
			'threshold'        => get_option( 'eatmyspam_threshold', 5 )
		) );

		if ( $result !== null ) {

			if ( isset( $result->spam ) && $result->spam === true ) {

				if ( get_option( 'eatmyspam_remove_spam' ) === 'on' ) {

					wp_safe_redirect( esc_url_raw( get_permalink( $post ) ) );
					die();


				} else {

					add_filter(
						'pre_comment_approved',
						create_function(
							'',
							'return "spam";'
						)
					);

				}

				if ( get_option( 'eatmyspam_send_notifications' ) === 'on' ) {

					add_filter(
						'comment_post',
						array(
							__CLASS__,
							'send_mail_notification'
						)
					);

				}
			}
		}

		return $comment;
	}

	/**
	 * @param int $id
	 *
	 * @return int
	 */
	public function send_mail_notification( $id ) {
		$comment = get_comment( $id, ARRAY_A );

		if ( empty( $comment ) ) {
			return $id;
		}

		if ( ! $post = get_post( $comment['comment_post_ID'] ) ) {
			return $id;
		}

		$subject = sprintf(
			'[%s] %s',
			stripslashes_deep(
				html_entity_decode(
					get_bloginfo( 'name' ),
					ENT_QUOTES
				)
			),
			esc_html__( 'Comment marked as spam', 'eat-my-spam' )
		);

		if ( ! $content = strip_tags( stripslashes( $comment['comment_content'] ) ) ) {
			$content = sprintf(
				'-- %s --',
				esc_html__( 'Content removed by EatMySpam', 'eat-my-spam' )
			);
		}

		$mail   = array();
		$mail[] = sprintf( __( 'A new comment on the post "%s" was marked as spam by EatMySpam.', 'eat-my-spam' ), strip_tags( $post->post_title ) );
		$mail[] = '';
		$mail[] = sprintf( '%s: %s', esc_html__( 'Author', 'eat-my-spam' ), $comment['comment_author'] );
		$mail[] = sprintf( '%s: %s', esc_html__( 'Email', 'eat-my-spam' ), $comment['comment_author_email'] );
		$mail[] = sprintf( '%s: %s', esc_html__( 'URL', 'eat-my-spam' ), $comment['comment_author_url'] );
		$mail[] = esc_html__( 'Comment:', 'eat-my-spam' );
		$mail[] = $content;
		$mail[] = '';
		$mail[] = sprintf(__( 'To manage comments, log in to WordPress at %s and navigate to: Comments > Spam', 'eat-my-spam' ), get_site_url());

		wp_mail(
			get_bloginfo( 'admin_email' ),
			$subject,
			join( chr( 10 ), $mail )
		);

		return $id;
	}

	/**
	 * make a post request to the eat my spam api
	 *
	 * @param string $method method to be called
	 * @param string $data array of data to be posted
	 *
	 * @return bool|object decoded response from eat my spam api
	 */
	protected function do_post( $method, $data ) {
		$url = 'https://' . self::API_HOST . '/' . $method;

		$args = array(
			'headers'     => array(
				'Content-Type' => 'application/json; charset=' . get_option( 'blog_charset' ),
				'User-Agent'   => 'EatMySpam/' . self::VERSION . ', WordPress/' . $GLOBALS['wp_version']
			),
			'body'        => json_encode( $data ),
			'httpversion' => '1.0',
			'timeout'     => 15
		);

		$response = wp_remote_post( $url, $args );

		if ( is_wp_error( $response ) ) {
			return false;
		} else {
			return json_decode( $response['body'] );
		}
	}

}

new EatMySpam;


eat-my-spam/classes/class.eat-my-spam.admin.php
<?php

/**
 * Security, checks if WordPress is running
 **/
if ( ! function_exists( 'add_action' ) ) :
	header( 'Status: 403 Forbidden' );
	header( 'HTTP/1.1 403 Forbidden' );
	exit();
endif;


/**
 * EatMySpam Admin Class
 *
 * @package EatMySpam
 */
final class EatMySpam_Admin {

	const API_HOST = 'api.eat-my-spam.de';

	/**
	 * Version
	 *
	 * @var string
	 **/
	protected $version = '0.6.2';

	/**
	 * @var array
	 */
	protected $rulesets;

	/**
	 * Constructor
	 *
	 * @access public
	 * @author Daniel Jost
	 **/
	public function __construct() {

		add_action( 'admin_init', array( $this, 'admin_init' ) );
		add_action( 'admin_menu', array( $this, 'admin_menu' ) );

		add_action( 'spam_comment', array( $this, 'report_spam' ) );
		add_action( 'unspam_comment', array( $this, 'report_ham' ) );

		add_filter( 'plugin_action_links_eat-my-spam/eat-my-spam.php', array( $this, 'add_settings_link' ) );
	}

	public function admin_init() {

		register_setting( 'eat-my-spam-settings', 'eatmyspam_threshold' );
		register_setting( 'eat-my-spam-settings', 'eatmyspam_remove_spam' );
		register_setting( 'eat-my-spam-settings', 'eatmyspam_send_notifications' );
		register_setting( 'eat-my-spam-settings', 'eatmyspam_excluded_rulesets' );
		register_setting( 'eat-my-spam-settings', 'eatmyspam_disable_reports' );

	}

	public function admin_menu() {

		add_options_page( 'EatMySpam Settings', 'EatMySpam', 'manage_options', 'eat-my-spam', array(
			$this,
			'menu_page_callback'
		) );

	}

	public function menu_page_callback() {

		$response = $this->load_rulesets();

		if ( $response === false || ! isset( $response->rulesets ) || ! is_array( $response->rulesets ) ) {
			die( 'error loading rulesets' );
		}

		$rulesets = $response->rulesets;

		$excludedRulesets = get_option( 'eatmyspam_excluded_rulesets', array() );

		if ( ! is_array( $excludedRulesets ) ) {
			$excludedRulesets = array();
		}

		include( plugin_dir_path( __FILE__ ) . '/../views/settings.php' );

	}

	protected function load_rulesets() {

		$url = 'https://' . self::API_HOST . '/rulesets';

		$args = array(
			'headers'     => array(
				'User-Agent' => 'EatMySpam/' . $this->version . ', WordPress/' . $GLOBALS['wp_version']
			),
			'httpversion' => '1.0',
			'timeout'     => 15
		);

		$response = wp_remote_get( $url, $args );

		if ( is_wp_error( $response ) ) {
			return false;
		} else {
			return json_decode( $response['body'] );
		}
	}

	public function add_settings_link( $links ) {

		$settings_link = '<a href="options-general.php?page=eat-my-spam">' . __( 'Settings', 'eat-my-spam' ) . '</a>';

		array_unshift( $links, $settings_link );

		return $links;
	}

	public function report_ham( $comment_id ) {
		if ( get_option( 'eatmyspam_disable_reports' ) !== 'on' ) {
			$this->report( 'ham', $comment_id );
		}
	}

	public function report_spam( $comment_id ) {
		if ( get_option( 'eatmyspam_disable_reports' ) !== 'on' ) {
			$this->report( 'spam', $comment_id );
		}
	}

	protected function report( $type, $comment_id ) {

		global $wpdb;

		$comment = $wpdb->get_row( $wpdb->prepare( "SELECT * FROM {$wpdb->comments} WHERE comment_ID = %d", $comment_id ), ARRAY_A );

		if ( $comment ) {

			$url = 'https://' . self::API_HOST . '/report';

			$data = array(
				'type'    => $type,
				'message' => $comment['comment_content']
			);

			$args = array(
				'headers'     => array(
					'Content-Type' => 'application/json; charset=' . get_option( 'blog_charset' ),
					'User-Agent'   => 'EatMySpam/' . $this->version . ', WordPress/' . $GLOBALS['wp_version']
				),
				'body'        => json_encode( $data ),
				'httpversion' => '1.0',
				'timeout'     => 15
			);

			$response = wp_remote_post( $url, $args );

			if ( is_wp_error( $response ) ) {
				return false;
			} else {
				return json_decode( $response['body'] );
			}

		}
	}

}

new EatMySpam_Admin();


eat-my-spam/includes/eat-my-spam.admin.php
<?php

/**
 * Security, checks if WordPress is running
 **/
if ( ! function_exists( 'add_action' ) ) :
	header( 'Status: 403 Forbidden' );
	header( 'HTTP/1.1 403 Forbidden' );
	exit();
endif;


/**
 * EatMySpam Admin Class
 *
 * @package EatMySpam
 */
class EatMySpam_Admin {

	const API_HOST = 'api.eat-my-spam.de';

	/**
	 * Version
	 *
	 * @var string
	 **/
	protected static $version = '0.8.1';

	/**
	 * @var array
	 */
	protected static $rulesets;

	/**
	 * Constructor
	 *
	 * @access public
	 * @author Daniel Jost
	 **/
	public static function init() {

		add_action( 'admin_init', array( __CLASS__, 'admin_init' ) );
		add_action( 'admin_menu', array( __CLASS__, 'admin_menu' ) );

		add_action( 'spam_comment', array( __CLASS__, 'report_spam' ) );
		add_action( 'unspam_comment', array( __CLASS__, 'report_ham' ) );

		add_filter( 'plugin_action_links_eat-my-spam/eat-my-spam.php', array( __CLASS__, 'add_settings_link' ) );

		self::init_plugin_sources();

		wp_enqueue_script('ems_script');
		wp_enqueue_script('ems_script2');
		wp_enqueue_style('ems_style');

	}

	public static function admin_init() {

		register_setting( 'eat-my-spam-settings', 'eatmyspam_threshold' );
		register_setting( 'eat-my-spam-settings', 'eatmyspam_remove_spam' );
		register_setting( 'eat-my-spam-settings', 'eatmyspam_send_notifications' );
		register_setting( 'eat-my-spam-settings', 'eatmyspam_excluded_rulesets' );
		register_setting( 'eat-my-spam-settings', 'eatmyspam_disable_reports' );
		register_setting( 'eat-my-spam-settings', 'eatmyspam_disable_cf7_integration' );
		register_setting( 'eat-my-spam-settings', 'eatmyspam_delete_spam_after_days' );
		register_setting( 'eat-my-spam-settings', 'eatmyspam_allowed_languages' );
		register_setting( 'eat-my-spam-settings', 'eatmyspam_disallowed_languages' );

	}

	public static function admin_menu() {

		$page = add_options_page( 'EatMySpam Settings', 'EatMySpam', 'manage_options', 'eat-my-spam', array(
			__CLASS__,
			'menu_page_callback'
		) );

		# integrate js and styles
		add_action( 'admin_print_scripts-' . $page, array( __CLASS__, 'add_options_script' ) );
		add_action( 'admin_print_styles-' . $page, array( __CLASS__, 'add_options_style' ) );
	}

	public static function menu_page_callback() {

		$response = self::load_rulesets();

		if ( $response === false || ! isset( $response->rulesets ) || ! is_array( $response->rulesets ) ) {
			die( 'error loading rulesets' );
		}

		$rulesets = $response->rulesets;

		$response = self::load_languages();

		if ( $response === false || ! isset( $response->languages ) || ! is_array( $response->languages ) ) {
			die( 'error loading languages' );
		}

		$languages = $response->languages;

		$excludedRulesets = get_option( 'eatmyspam_excluded_rulesets', array() );

		if ( ! is_array( $excludedRulesets ) ) {
			$excludedRulesets = array();
		}

		$allowedLanguages = get_option( 'eatmyspam_allowed_languages', array() );

		if ( ! is_array( $allowedLanguages ) ) {
			$allowedLanguages = array();
		}

		$disallowedLanguages = get_option( 'eatmyspam_disallowed_languages', array() );

		if ( ! is_array( $disallowedLanguages ) ) {
			$disallowedLanguages = array();
		}

		include( plugin_dir_path( __FILE__ ) . '/../views/settings.php' );

	}

	protected static function load_rulesets() {

		$url = 'https://' . self::API_HOST . '/rulesets';

		$args = array(
			'headers'     => array(
				'User-Agent' => 'EatMySpam/' . self::$version . ', WordPress/' . $GLOBALS['wp_version']
			),
			'httpversion' => '1.0',
			'timeout'     => 5
		);

		$response = wp_remote_get( $url, $args );

		if ( is_wp_error( $response ) ) {
			return false;
		} else {
			return json_decode( $response['body'] );
		}
	}

	protected static function load_languages() {

		$url = 'https://' . self::API_HOST . '/languages';

		$args = array(
			'headers'     => array(
				'User-Agent' => 'EatMySpam/' . self::$version . ', WordPress/' . $GLOBALS['wp_version']
			),
			'httpversion' => '1.0',
			'timeout'     => 5
		);

		$response = wp_remote_get( $url, $args );

		if ( is_wp_error( $response ) ) {
			return false;
		} else {
			return json_decode( $response['body'] );
		}
	}

	public static function add_settings_link( $links ) {

		$settings_link = '<a href="options-general.php?page=eat-my-spam">' . __( 'Settings', 'eat-my-spam' ) . '</a>';

		array_unshift( $links, $settings_link );

		return $links;
	}

	public static function report_ham( $comment_id ) {
		if ( get_option( 'eatmyspam_disable_reports' ) !== 'on' ) {
			self::report( 'ham', $comment_id );
		}
	}

	public function report_spam( $comment_id ) {
		if ( get_option( 'eatmyspam_disable_reports' ) !== 'on' ) {
			self::report( 'spam', $comment_id );
		}
	}

	protected static function report( $type, $comment_id ) {

		global $wpdb;

		$comment = $wpdb->get_row( $wpdb->prepare( "SELECT * FROM {$wpdb->comments} WHERE comment_ID = %d", $comment_id ), ARRAY_A );

		if ( $comment ) {

			$url = 'https://' . self::API_HOST . '/report';

			$data = array(
				'type'    => $type,
				'message' => $comment['comment_content']
			);

			$args = array(
				'headers'     => array(
					'Content-Type' => 'application/json; charset=' . get_option( 'blog_charset' ),
					'User-Agent'   => 'EatMySpam/' . self::$version . ', WordPress/' . $GLOBALS['wp_version']
				),
				'body'        => json_encode( $data ),
				'httpversion' => '1.0',
				'timeout'     => 5
			);

			$response = wp_remote_post( $url, $args );

			if ( is_wp_error( $response ) ) {
				return false;
			} else {
				return json_decode( $response['body'] );
			}

		}
	}

	public static function init_plugin_sources()
	{
		wp_register_script(
			'ems_script',
			plugins_url('../js/select2.full.min.js', __FILE__),
			array('jquery')
		);

		wp_register_script(
			'ems_script2',
			plugins_url('../js/main.js', __FILE__),
			array('ems_script')
		);

		wp_register_style(
			'ems_style',
			plugins_url('../css/select2.min.css', __FILE__),
			array()
		);
	}

}

EatMySpam_Admin::init();